### ISSUE TITLE
Caching directories between runs of a task

### ISSUE THREAD

**Issue by vito** [concourse/concourse#230]
Common use case: fetching dependencies, syncing BOSH blobs, etc.

Caching the directories that these fetch into and update would dramatically speed up a bunch of builds. Our own ATC JS build spends 99% of its time just downloading `npm` packages.

Proposal:

Add a `cache` field to task configs, like so:

``` yaml

---
platform: linux

inputs:
- name: my-release

cache:
- path: my-release/.blobs

run:
  path: my-release/ci/scripts/create-release
```

Then, given I have a pipeline like so:

``` yaml
jobs:
- name: make-release
  plan:
  - get: my-release
  - task: create-release
    file: my-release/ci/create-release.yml
```

This would cache the directory `my-release/.blobs` between runs of that specific task in its job's build plan. So, the cache lookup key would be something like `team-id+pipeline-id+job-name+task-name`.

Notes:
- This should also have the guarantee that two concurrent builds of the same job do not pollute each others' caches. There should be some sort of copy-on-write semantics, such that each job gets its own copy of the cache (initially empty), and at the end all other caches are marked "stale" and are set to expire.
- Assumes tools will be durable to the directory being initially present + empty on the initial cache run. I think this should be fine. Without this it'll be very annoying to orchestrate.
- The caching is for purely ephemeral data, so it doesn't sacrifice Concourse's "not being a source of truth" principle.
- Has the same cache warming semantics as `get`s, i.e. it may take a bit for the cache to warm across the workers; it does not influence worker placement.
- Has no effect on one-off builds, as there is not enough information to scope/correlate the caches (compared to a job build).

**Comment by xoebus** [comment]
:+1:

It's an important feature but I can see it getting abused. Maybe we need a best practices document that sets some guidelines for having a sane pipeline. e.g. don't publish a different artefact to the same version number otherwise nothing will work properly.

**Comment by nicorikken** [comment]
I'd certainly like to see this come about. Our use-case includes downloading NPM dependencies as well, but also downloading Nix-packages. Copy-on-write would suit that use-case.

Sharing this cache even among pipelines would make sense efficiency-wise, although that would introduce (semantic) dependencies between pipelines.

**Comment by vito** [comment]
@nicorikken We'll probably not start with cross-pipeline or even cross-job caching as that opens another can of worms: how to configure which caches apply to which tasks/etc. I could see that becoming a nightmare pretty quickly if you e.g. copy a job from another pipeline that has a cache name the same as yours because it was called something too generic.

So we'll probably just stick with as tightly-scoped as possible (individual task within job).

**Comment by nicorikken** [comment]
@vito Totally agree, makes most sense. There are other 'hacks' for getting around such issues, and worst-case somebody would have to write a specific resource targeting the local disk.

**Comment by jchester** [comment]
:+1: from me, both privately (I have a pipeline pulling umpteen zillion jars) and professionally (pivnet is a hungry gem consumer).

**Comment by devcurmudgeon** [comment]
not sure i understand this discussion properly, but i think just supporting the possibility to have 'cache' directory/ies untouched between jobs would be useful on its own. various tools have their own cache lookup processes (eg bitbake)

**Comment by tomwhoiscontrary** [comment]
👍 we've got microservices built with Gradle, so every one of our 27 tasks starts by downloading Gradle, then all our dependencies, and it's horrendous. Cache plox.

As a stop-gap, could we write some sort of pre-loading resource? It would somehow download the Gradle binary and some set of dependencies, then we would copy that into place before kicking off the build. That would let us exploit Concourse's local caching of resources, right?

**Comment by simonvanderveldt** [comment]
> @nicorikken We'll probably not start with cross-pipeline or even cross-job caching as that opens another can of worms: how to configure which caches apply to which tasks/etc. I could see that becoming a nightmare pretty quickly if you e.g. copy a job from another pipeline that has a cache name the same as yours because it was called something too generic.
> 
> So we'll probably just stick with as tightly-scoped as possible (individual task within job).

@vito While I do understand it will be hard to make sure people don't use it for the wrong reasons we see a serious speedup with shared caches between pipelines (and by extension also between jobs).
But the caches we share in these cases are the "normal" per user package manager's caches, so for example for npm this is [`~/.npm/](https://docs.npmjs.com/misc/config#cache) and for sbt this is `~/.ivy2/`. 

Would it be possible to support this somehow?


> @vito Totally agree, makes most sense. There are other 'hacks' for getting around such issues, and worst-case somebody would have to write a specific resource targeting the local disk.

@nicorikken Can you share those other "hacks"?

**Comment by seadowg** [comment]
As a wee push: not having this feature is still causing major problems with Concourse at some enterprise companies due to terrible connections (usually through a VPN). I've seen this drop down the backlog a couple of times so just thought I'd add my feelings from the world of self important anecdotal evidence.

**Comment by tomwhoiscontrary** [comment]
@seadowg Oh, i really should mail out about this: we made a Gradle-specific caching thingy on our project. Because Docker Hub automatic builds need public repos, i had to make it public, so i guess you can read it?

https://github.com/projectfalcon/gradle-cache-resource

You might be able to do something similar. The example pipeline builds Spring with it.

The way it works is that you still need to download all the dependencies in one go, but only once, after that it's cached. Until one of them changes, then you have to download them all again.

Also, i've been off sick, so by now Thorbs may have replaced this with some rival saucer science.

**Comment by nicorikken** [comment]
@simonvanderveldt The hacks were a while ago, and I hope Concourse has improved on many levels since then. We tried sharing the local directories, but that didn't go so well (don't remember). We eventually were having Dockers on Docker Hub in multiple stages in the pipeline: [Base build dependency image] -> (fetch latest dependencies) -> [Up-to-date build dependency image] -> (build) -> [finished build] -> (test) -> [build and tested image]

**Comment by poida** [comment]
We also need to be able to cache folders (NPM modules folder) between jobs.
Currently Concourse runs 2-3 times slower than similar pipeline on Jenkins just because of NPM INSTALL.
Any ideas, hacks to speed it up?

**Comment by chipx86** [comment]
For ours, we're using a fork of concourse-rsync-resource (https://github.com/beanbaginc/concourse-rsync-resource) and define a resource for each type of build we need to share. This looks like:

```yaml
- name: my-project-env
  type: rsync-resource
  check_every: 10m
  source:
      server: our-builds.example.com
      base_dir: /path/to/unique/build/dest
      user: build-username
      private_key: {{builds-ssh-key}}
```

We have a job for each of these that builds the environment and puts it:

```yaml
- name: my-project-prep
  plan:
      - task: prep
        file: tasks/my-project/prep.yaml
        input_mapping:
            source: my-project-src

      - put: my-project-env
        name:
             sync_dir: env
```

Jobs that need it just do a `get`:

```yaml
- get: my-project-env
  trigger: true
  passed:
      my-project-prep
```

That `prep.yaml` file looks like:

```yaml
platform: linux

image_resource:
    ...

inputs:
    ...

outputs:
    - name: env
      path: env

run:
    path: tasks/my-project/prep.sh
```

The `prep.sh` just does what it needs to do, and then outputs to wherever it needs in the `env` directory. Other tasks can then make use of what's in there.

For node packages, we use `npm-cache` to help with the cache. We have a helper function for use in the `prep.sh` scripts to set up that cache, install the modules, and store them in the environment:

```shell
function build_and_store_node_modules() {
    cache_dir=$BUILDS_DIR/env/.package_cache

    mkdir $cache_dir
    ln -s $cache_dir ~/.package_cache

    echo Installing node modules...
    npm install -g npm-cache
    npm-cache install
}
```

Scripts that need to load the environment then just call this helper:

```shell
function restore_node_modules() {
    echo Restoring node modules...
    ln -s $BUILDS_DIR/env/.package_cache ~
}
```

Hope that helps! I can provide more examples if needed.

**Comment by tomwhoiscontrary** [comment]
@poida You should be able to use the same approach with NPM that we used with Gradle - wrap the Git resource in [a script which pulls your code, downloads the dependencies, then exposes the dependencies as a resource](https://github.com/projectfalcon/gradle-cache-resource/blob/master/in). Replace `build.gradle` with `package.json`, `./gradlew __cacheDependencies` with `npm install`, and `~/.gradle` with `~/.node_modules`.

**Comment by jchester** [comment]
I gave this more thought recently and have changed my position. I still see the need, but I think it should be handled as a new resource.

The key implementation problem seems to be that any cache will either need to be centralised, or consistent amongst workers. 

I bring this up because my first thought was: "why not write a resource that punches a hole through to a directory on the worker FS?", followed by the second: "uh, what if there is more than one worker, Jacques?"

I think that adopting the lightest possible blob store as a core-supported resource is the go here. It would retain existing Concourse semantics, would avoid changes to core logic and would provide a relatively smooth pathway for folk switching to and from remote blobstores as the case may be.

I don't feel qualified to evaluate the alternatives, but it seems to me that something that can distribute itself less-than-stupidly across workers would be the goal.

Some quick googling: The trendiest right now is minio (Golang u guise!1!), but it's not distributed yet. SX Cluster is distributed and more mature, but mired under a GPL cloud. Ceph is very mature and under LGPL, not sure how heavyweight it is. Ambry comes from LinkedIn and is presumably pretty robust, but requires a JVM.

**Comment by poida** [comment]
@jchester is what you're suggesting what baggage claim already does for concourse volumes? I wonder if that could be leveraged somehow into a resource type.

**Comment by dougtweedy** [comment]
+1 having this for our gradle builds is essential for caching dependencies. Using volume mounts on the .gradle directory would work too, but the option needs to be exposed to the concourse tooling.

**Comment by aarondl** [comment]
+1 Building a caching mechanism (we use an rsync server with the rsync resource) is silly and is painful because all of concourses paradigms make it hard to keep things between builds (like not being able to use the same directory for put/get).

**Comment by poida** [comment]
Thanks @tomwhoiscontrary, what sort of speed up do you get with your gradle cache-resource to your pipeline? Did you have to tweak concourse settings to keep the resources around? Does concourse evict the resources out of the cache often?

**Comment by tomwhoiscontrary** [comment]
@poida The speedup appears to be equal to the time taken to download the resources, so a minute or so. I've never noticed the resource being rebuilt after being discarded from the cache. But then, ours is a small pipeline, with only four resources that actually produce files, so there is not a lot of cache pressure. A small problem is gratuitous rebuilds - if you make an irrelevant tweak to a build file (eg changing whitespace), the resource will rebuild. That's still better than rebuilding every time, though.

**Comment by jchester** [comment]
I made [a stab](https://github.com/jchesterpivotal/concourse-with-minio-deployment) at setting up a resource during a hackday about a month ago. Didn't get a 10th of what I wanted, but I at least demonstrated using a co-deployed blobstore to avoid having to hit the internet.

**Comment by savagematt** [comment]
This is the only thing that's stopping us using Concourse.

Aside from having a LOT of data in our gradle/node/maven/whatever caches, we also use docker-compose to run builds against local, isolated databases. Our proof-of-concept needs to download these images from the registry on every build.

The end result is that builds run one or two orders of magnitude more slowly in Concourse than they do on our GOCD server (which we are trying very hard to migrate away from).

At the moment, tasks in GOCD mount the `~/.m2`, `~/.npm`, `~/.every_other_build_tool` directories from the agent into a docker container that we use to run the build. Thanks to only running one task at a time per agent, this works out fine.

One concern with having one cache per task would be that these caches can get pretty big and we have a LOT of pipelines. Although storage is cheap, of course.

Is there any rough indication of when this feature might land? We need to make a call between waiting for Concourse and investing in exploring other options (which are less attractive to us for many other reasons).

**Comment by jemc** [comment]
Like @savagematt, we use `docker-compose` extensively in our test suite (running a docker-in-docker container), and downloading/building images takes the vast majority of our build time for every task.

Even just being able to share docker cached image layers between tasks that *happen* to be running on the same worker would be a godsend.

**Comment by jemc** [comment]
There seems to be a lot of talk here about the problems revolving around centralization - that is, needing to centralize the cache so that it is available from all possible concourse workers.

I think that's looking at it the wrong way.

Caching should be a speedup to doing repetitive work, and all operations that use a cache should be capable of "falling back" and doing the operation to fetch/build what's needed if the cached result is not available.  In other words, tasks which use a cache *shouldn't take that cache for granted* - they should be able to recreate the parts of the cache they touch if necessary.

With that mindset, there's no need for a centralized cache at all - each worker could have its own local cache, which gets populated by the results of tasks that get allocated and run there.  If the cached result you're looking for doesn't exist on your local worker, or maybe the entire cache was wiped recently?  No problem - you just proceed to build/fetch what you need, and fill the cache so that the next task looking for that result doesn't have to do the same work over again.

**Comment by tomwhoiscontrary** [comment]
If you want docker images in builds, can you get them from a docker image resource? Those get cached. You might need to do a bit of scripting to move the images from the resource directories to a place where docker-compose will find them, but who doesn't love a bit of scripting?

**Comment by jemc** [comment]
@tomwhoiscontrary - that's more or less what we're looking at doing in one part of our CI, but it's a lot of hassle to do something that we'd get "for free" if the docker-in-docker tasks could share the same docker daemon.

Also, if I understand correctly the docker image resource requires being linked to a centralized registry, which is not as resilient as the kind of decentralized approach I discussed above, and setting up/maintaining a private registry just for this purpose is kind of a hassle.  We already have a private registry for our production images, but we wouldn't want to clutter it up with a bunch of CI intermediate images that we want to wipe on a more regular basis.

**Comment by DenisIzmaylov** [comment]
We're really waiting to get this feature. At the moment our pipelines build time looks very terrible. We have made a few "hacks" to improve that but it added a unnecessary complexity to the pipeline.

Definitely the resources should be cached. GitLab CI should not be faster as we have right now. We love Concourse and we are looking forward to speed up our builds.

**Comment by analytically** [comment]
See https://github.com/ymedlop/npm-cache-resource/

**Comment by jshearer** [comment]
My project requires a bunch of pip dependencies (similar to gems, etc) being installed to work. I would love to just have to download these once, and update them when the dependencies are updated later. Using the proposed `cache` option would be perfect for me. Are there plans for this to be implemented in the future?

**Comment by Rukenshia** [comment]
I'd also love to have some way of caching these. I'm building a Rust application and every build takes 10 minutes because cargo has to redownload and compile each dependency again. Otherwise the build would take less than a minute.

**Comment by seadowg** [comment]
@vito any plans to prioritise this higher? It's fallen down to May this year I see. Still causing problems for teams attempting to use Concourse on not super fast internet connections.

**Comment by timrchavez** [comment]
:+1: Having this would win over many developers where I work. It's one of the loudest complaints we have.

**Comment by JohannesRudolph** [comment]
We've had this same issue for a long time and it was very annoying until we came up with a non-intrusive caching strategy using a Task and two short scripts. I've wrapped it up as a sample repo here: https://github.com/Meshcloud/concourse-cached-pipeline

In any case, I'm curious about @vito 's plans to add this natively to Concourse and how and when we can expect this to happen.

**Comment by fralalonde** [comment]
The cache needs to be per-worker to maximize locality, accessible concurrently and totally transparent to the jobs. I'm thinking of running a caching proxy on every worker node, and route any request to our Artifactory server through it. The jobs will still "download" every dependency on each run, but transfers will be from the local worker proxy and thus will never hit the wire after first access. 

Also, I don't know about other build systems, but Maven with [Takari extensions](http://takari.io/book/30-team-maven.html#concurrent-safe-local-repository) allows .m2/repository to safely be updated by two concurrent build processes. This would make it possible for two jobs to use the same cache as a local network mount, or a shared docker volume, which would not require a proxy. If concourse could provide a default way to access worker-local storage to any / every job... Maybe this already exists? I've been out of the loop for a while.

**Comment by vito** [comment]
Some investigation needed in to parallel execution of builds with the same cache. But we'll pick it up and figure it out as we go.

Worst-case ontario, we could just not care, and say if you've got parallel hits of the same cache, you can just set `serial: true` if there are problems.

**Comment by pn-santos** [comment]
One thing I don't quite understand from the proposed solution, how would the cache be evicted/reset? 

Would that be something that the task is responsible for? Could fly for example offer a way to evict caches of a certain job?

Also regarding the scope of caches, won't making them task scoped make the size requirements for each worker go up considerably? i.e it I have 3 tasks that use the same cache I'll have 3 copies of the same thing if I cache it on all 3 tasks right? Eventually that data would be present in all workers...

Wouldn't it be preferable to make them per job? (so making the key `team-id+pipeline-id+job-name`). It would be like a task input that could also be an output (i.e. like mutating data in a task input path that is visible to downstream tasks). If one considers the more common scenarios (npm, maven/gradle, go vendor, etc) its very likely that these caches will be re-used by tasks within a job. Handling concurrency within a job is not that hard (vs handling cross pipeline).

**Comment by mariash** [comment]
The initial approach is to introduce `worker_task_caches` table with the following fields:
- worker_name
- job_id
- step_name (is step name unique within the job? or you can have several with the same name?)

Plan ID is not consistent, it is just an incremented number within pipeline (see NewPlanFactory). It will be different for each build.

Volumes can now belong to worker_task_caches. Same way as to worker_resource_caches.

This will allow to define clear garbage collection for those volumes: 
- Delete worker_task_cache when it is no longer needed, e.g. it was removed from pipeline.
- This will nullify worker_task_cache_id in volumes and that volume will become a candidate for garbage collection.

Flow:

When task is running:
  For each specified cache in task:
    If cache volume exists:
    - Mount a COW volume at the specified path. COW volume will belong to container.
    If cache volume does not exist:
    - Mount a new volume at the specified path. COW volume will belong to container.

After task is run:
  For each specified cache in task:
    If it was a COW volume and sha is different from parent's volume:
    - Create an import volume from COW volume
    - Set worker_task_cache_id on an import volume and remove worker_task_cache_id from parent volume (in transaction). That will make old parent volume a candidate for garbage collection.
    ** If two tasks modified COW volume in parallel, figure out which one to pick. E.g. highest build ID.

TODO:
- [ ] can we rely on `step_name`? if not figure out how to uniquely identify step within job.
- [ ] rename `cache` to `caches` in task config? to be consistent with inputs/outputs?
- [ ] garbage collect task caches
- [ ] concurrency: when two builds are running in parallel which volume will be promoted to task cache?
- [ ] optimization: as a first iteration import volume will be created every time. Add optimization to check if it was a new volume or sha did not change

**Comment by vito** [comment]
@mariash minor thing - noticed these show up as "unknown" in `fly volumes`. could we show something special for these? e.g. `task-cache` and maybe the pipeline/job/step?

**Comment by vito** [comment]
Breaking that last bit out as a separate issue in 3.3.x since it's not critical enough to block release.

**Comment by alnavart** [comment]
Hi, based on last @mariash comment.  Where reflect that "garbage collect task caches" was not done. Are cache volumes never removed by the garbage collector?
We had some issues related with the disk usage after enable some caches and can be related with it.
Thanks
